### PR TITLE
Run all glob tests against chokidar too

### PR DESCRIPTION
**Pull request by aomarks**
This PR runs every glob test against chokidar, since we should have identical behavior in theory. We currently don't, so there are a lot of tests that get excluded for the chokidar case.

Tracking the task of making them equivalent at https://github.com/google/wireit/issues/550, but figured I'd land the tests first so that we at least don't regress, and have something easy to track for improvements in future.

### CODE REVIEW

File: src/test/glob.test.ts [hunk 1]
```diff
@@ -76,29 +80,60 @@ test.before.each(async (ctx) => {
...
+      } else if (mode === 'watch') {
+        const actual: string[] = [];
+        if (patterns.length > 0) {
+          const {watcher} = makeWatcher(
+            patterns,
+            rig.resolve(cwd),
+            () => undefined,
+            // We need ignoreInitial=false because we need the initial "add"
+            // events to find out what chokidar has found (we usually only care
+            // about changes, not intial files).
```
Comment (AndrewJakubowicz) [inline src/test/glob.test.ts:118]: Nit: `intial` -> `initial`.

File: src/test/glob.test.ts [hunk 2]
```diff
@@ -120,491 +155,584 @@ test.after.each(async (ctx) => {
...
+      ],
+      patterns: ['symlink'],
+      expected: ['symlink'],
+    }));
+
+  test(`[${mode}] explicit directory excluded when includeDirectories=false`, ({
+    check,
+  }) => check({mode, files: ['foo/'], patterns: ['foo'], expected: []}));
+
+  if (mode !== 'watch') {
```
Comment (AndrewJakubowicz) [inline src/test/glob.test.ts:254]: Nit: Could abstract this check to be more readable. E.g. `skipWatchTest()', which could have a comment.
Optional.

File: src/test/glob.test.ts [hunk 3]
```diff
@@ -76,29 +80,60 @@ test.before.each(async (ctx) => {
...
+        } else if (error !== undefined) {
+          throw error;
+        } else if (actual === undefined) {
+          throw new Error('Actual was undefined');
+        } else {
+          const actualPaths = actual.map((file) => file.path);
+          assert.equal(actualPaths.sort(), expected.sort());
+        }
+      } else if (mode === 'watch') {
+        const actual: string[] = [];
```
Comment (AndrewJakubowicz) [inline src/test/glob.test.ts:110]: Nit: Could add a symmetrical comment here with a similar sentiment to "`// One-time globbing (via fast-glob).`.

**Review by AndrewJakubowicz — APPROVED** [review]
Nice! Non blocking nits.

File: src/test/glob.test.ts (hunk 3 above)
Comment (aomarks) [reply src/test/glob.test.ts:109]: Removed the other comment, since it's pretty much self-describing

File: src/test/glob.test.ts (hunk 2 above)
Comment (aomarks) [reply src/test/glob.test.ts]: Done!

File: src/test/glob.test.ts (hunk 1 above)
Comment (aomarks) [reply src/test/glob.test.ts:117]: Done

**Comment by aomarks** [comment]
Sorry for the churn, I just force-pushed a much more incremental version of this PR that is easier to read.

Also switched to using the `test.skip` function, so that it's more obvious in the output how many are being skipped.

File: src/test/glob.test.ts [hunk 4]
```diff
@@ -120,491 +154,646 @@ test.after.each(async (ctx) => {
...
-    followSymlinks: true,
-    includeDirectories: true,
-    expandDirectories: false,
-    throwIfOutsideCwd: false,
+for (const mode of ['once', 'watch'] as const) {
+  // eslint-disable-next-line @typescript-eslint/unbound-method
+  const skipIfWatch = mode === 'watch' ? test.skip : test;
+  const skipIfWatchOnWindows =
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    mode === 'watch' && IS_WINDOWS ? test.skip : test;
```
Comment (AndrewJakubowicz) [inline src/test/glob.test.ts:162]: Nice!

File: src/watcher.ts [hunk 5]
```diff
@@ -448,22 +448,30 @@ const watchPathsEqual = (
...
+ * @param ignoreInitial Ignore the initial "add" events emitted when chokidar
+ * first discovers each file. We already do an initial run, so these events are
+ * just noise that may trigger an unnecessary second run.
+ * https://github.com/paulmillr/chokidar#path-filtering
+ */
+export const makeWatcher = (
   patterns: string[],
   cwd: string,
-  callback: () => void
+  callback: () => void,
```
Comment (AndrewJakubowicz) [inline src/watcher.ts:462]: This is unrelated, but what does this callback do?

**Review by AndrewJakubowicz — APPROVED** [review]
Looks great!

File: src/test/glob.test.ts (hunk 4 above)
Comment (AndrewJakubowicz) [reply src/test/glob.test.ts:162]: This `test.skip : test` technique is great!

File: src/watcher.ts (hunk 5 above)
Comment (aomarks) [reply src/watcher.ts:462]: It just gets passed to the `all` callback

**Pull request merged**